### PR TITLE
TUN-1.7: Tunnel Interfaces disable and enable - Interface Based GRE Tunnel

### LINKED ISSUE #1604: TUN-1.7: Tunnel Interfaces disable and enable - Interface Based GRE Tunnel

### PR DESCRIPTION
**Pull request by SenthilkumarSelvaganapathy**
Configuration
Configure DUT with 8 GRE encapsulation tunnels and configure another router with 8 GRE Decapsulation tunnel interfaces.
Configure 4 tunnel as IPv4 tunnel source and destination address , 4 as IPv6 tunnel source and destination address
Configure static router to point original destination to 8 tunnel interface to do overlay load balance
Keep topology tunnel destination will be reachable via 2 underlay interface on both routers
Send IPv4 flow and IPv6 flow and validate tunnel load balance and physical interface load balance
Disable and enable  5 numbers of Tunnel interface being used  One by one      
Enable or bring up the same interfaces which were disabled in previous step by One by one
       

Validation Coverage:
Static route using the disable interface should become invalid and should not be used in forwarding traffic
Incoming traffic on DUT-PORT1 should be load balanced to available Tunnel interfaces for encapsulation
Incoming traffic flow should be equally distributed for Encapsulation(ECMP)
No traffic loss expected
Disabling and enabling the number of tunnel interfaces and related static route shouldn’t case traffic drops
Verify the Next hop counters for packet being diverted or sent for encapsulation
Verify the tunnel interfaces counters to confirm the traffic encapsulation
Verify the tunnel interfaces traffic/flow for equal distribution for optimal load balancing
After decapsulation, traffic should be load balanced/hash to all available L3 ECMP or LAG or combination of both 
This resolves #1604

### CODE REVIEW

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/README.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      *  All 5 at once
+ *   Incoming traffic on DUT-PORT1 should start using additional Tunnel Interfaces  for encapsulation
+ *   Incoming traffic flow should be equally distributed for Encapsulation(ECMP)
+ *   No traffic loss expected
+ *   Disabling and enabling the number of tunnel interfaces and related static route shouldn’t case traffic drops
+ *   Verify the Next hop counters for packet being diverted or sent for encapsulation
+ *   Verify the tunnel interfaces counters to confirm the traffic encapsulation
+ *   Verify the tunnel interfaces traffic/flow for equal distribution for optimal load balancing
+ *   After decapsulation, traffic should be load balanced/hash to all available L3 ECMP or LAG or combination of both features
+ *   Verify the tunnel interfaces counters to confirm the traffic decapsulation
```
Comment (sachendras) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/README.md:32]: Some more additional steps

Validate system for:
 - Health-1.1
 - No feature related error or drop counters incrementing, discussion with vendors required to highlight additional fields to monitor based on implementation and architecture

Comment (SenthilkumarSelvaganapathy) [reply feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/README.md]: Hi sachendras,

I have corrected "README.md" file as per the Testplan
Also I have corrected "CONTRIBUTING.md" error.

Regards,
Senthil Kumar

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/README.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+ *   Incoming traffic flow should be equally distributed for Encapsulation(ECMP) 
+ *   No traffic loss expected 
+ *   Enable the tunnel interfaces:  
+     *  Enable or bring up the same interfaces which were disabled in previous step 
+        *  One by one 
+        *  All 5 at once 
+ *   Incoming traffic on DUT-PORT1 should start using additional Tunnel Interfaces  for encapsulation 
+ *   Incoming traffic flow should be equally distributed for Encapsulation(ECMP) 
+ *   No traffic loss expected 
+ *   Disabling and enabling the number of tunnel interfaces and related static route shouldn’t case traffic drops
```
Comment (sachendras) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/README.md:25]:
```suggestion
 *   Disabling and enabling the number of tunnel interfaces and related static route shouldn’t cause traffic drops 
```

**Review by sachendras — DISMISSED** [review]

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/README.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
+
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/README.md:1]: delete blank line

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/README.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+ ##  TUN-1.7: Tunnel Interfaces disable and enable - Interface Based GRE Tunnel
+ ## Summary
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/README.md:3]: Add a blank line before the Summary and Procedure headers for consistency with other files

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/README.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+ *   state/counters/in-discarded-pkts 
+ *   state/counters/out-error-pkts 
+ *   state/counters/out-forwarded-pkts 
+ *   state/counters/out-forwarded-octets 
+ *   state/counters/out-discarded-pkts 
+
+ ## Topology:
+ *   otg:port1 <--> port1:dut1:port3 <--> port3:dut2:port5<--->otg:port5
+ *   otg:port2 <--> port2:dut1:port4 <--> port4:dut2:port6<--->otg:port6
+
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/README.md:71]: delete trailing blank line

File: internal/deviations/deviations.go [hunk 6]
```diff
@@ -597,14 +597,14 @@ func ISISCounterPartChangesUnsupported(dut *ondatra.DUTDevice) bool {
...
-	return lookupDUTDeviations(dut).GetTransceiverThresholdsUnsupported()
+// TunnelStatePathUnsupported returns true for devices that require configuring
+// /interfaces/interface/state/counters/in-pkts, in-octets,out-pkts, out-octetsis not supported.
+func TunnelStatePathUnsupported(dut *ondatra.DUTDevice) bool {
+	return lookupDUTDeviations(dut).GetTunnelStatePathUnsupported()
 }
 
-// InterfaceLoopbackModeRawGnmi returns true if interface loopback mode needs to be updated using raw gnmi API due to server version.
-// Default value is false.
-func InterfaceLoopbackModeRawGnmi(dut *ondatra.DUTDevice) bool {
```
Comment (greg-dennis) [inline internal/deviations/deviations.go:608]: What do you want to do about the existing test using this deviation?
https://github.com/openconfig/featureprofiles/blob/24d10056e7453a2750439d7a329a4c02ba838616/feature/experimental/interface/interface_loopback_aggregate/otg_tests/interface_loopback_aggregate/interface_loopback_aggregate_test.go#L231

File: internal/deviations/deviations.go [hunk 7]
```diff
@@ -597,14 +597,14 @@ func ISISCounterPartChangesUnsupported(dut *ondatra.DUTDevice) bool {
 	return lookupDUTDeviations(dut).GetIsisCounterPartChangesUnsupported()
 }
 
-// TransceiverThresholdsUnsupported returns true if the device does not support threshold container under /components/component/transceiver.
-// Default value is false.
-func TransceiverThresholdsUnsupported(dut *ondatra.DUTDevice) bool {
```
Comment (greg-dennis) [inline internal/deviations/deviations.go:602]: What do you want to do about the existing test using this deviation?
https://github.com/openconfig/featureprofiles/blob/24d10056e7453a2750439d7a329a4c02ba838616/feature/platform/tests/optics_power_and_bias_current_test/optics_power_and_bias_current_test.go#L97

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 8]
```diff
@@ -0,0 +1,1558 @@
+// Copyright 2022 Google LLC
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:1]: 2023

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 9]
```diff
@@ -0,0 +1,1558 @@
...
+	flow1           string
+	flow2           string
+	flow3           string
+	flow4           string
+	trafficDuration int64
+	trafficRate     int64
+}
+
+func GetNetworkAddress(t *testing.T, address string, mask int) string {
+
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:206]: for consistency with the rest of the codebase, delete blank lines at the beginning of your methods

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 10]
```diff
@@ -0,0 +1,1558 @@
...
+}
+
+func configIPv4StaticRoute(t *testing.T, dut *ondatra.DUTDevice, prefix string, nexthop string, index string) {
+	ni := oc.NetworkInstance{Name: ygot.String(deviations.DefaultNetworkInstance(dut))}
+	static := ni.GetOrCreateProtocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_STATIC, deviations.StaticProtocolName(dut))
+	sr := static.GetOrCreateStatic(prefix)
+	nh := sr.GetOrCreateNextHop(index)
+	nh.NextHop = oc.UnionString(nexthop)
+	gnmi.Update(t, dut, gnmi.OC().NetworkInstance(deviations.DefaultNetworkInstance(dut)).Protocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_STATIC, deviations.StaticProtocolName(dut)).Config(), static)
+
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:634]: delete this and any other blank line at the end of functions

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 11]
```diff
@@ -0,0 +1,1558 @@
...
+	r1Lo0Ut3Ipv6Add string
+	flow1           string
+	flow2           string
+	flow3           string
+	flow4           string
+	trafficDuration int64
+	trafficRate     int64
+}
+
+func GetNetworkAddress(t *testing.T, address string, mask int) string {
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:205]: Go style guide says functions shouldn't start with "Get". "Fetch" or "Lookup" here is fine

Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:205]: no need to export this function or any of the other helper functions in this file. Make them lowercase

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 12]
```diff
@@ -0,0 +1,1558 @@
...
+
+	t.Log("gnmiClient Set CLI config")
+	if _, err = gnmiClient.Set(context.Background(), gpbSetRequest); err != nil {
+		t.Fatalf("gnmiClient.Set() with unexpected error: %v", err)
+	}
+}
+
+func TestFtiTunnels(t *testing.T) {
+
+	p := &parameters{
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:740]: This massive parameters type looks unnecessary. Just define them as constants local to this test function, and feel free to inline the ones that are used only once

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 13]
```diff
@@ -0,0 +1,1558 @@
...
+		ipv6FullMask:    128,
+		flow1:           "IPv4-flow1",
+		flow2:           "IPv4-flow2",
+		flow3:           "IPv6-flow3",
+		flow4:           "IPv6-flow4",
+		trafficDuration: 60,
+		trafficRate:     1000,
+	}
+	t.Logf("the input variable %+v", p)
+	t.Helper()
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:829]: This isn't a helper function -- it's a top-level test function, so drop this line

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 14]
```diff
@@ -0,0 +1,1558 @@
...
+
+	dut1 := ondatra.DUT(t, "dut1")
+	dut1Intf1 := dut1.Port(t, "port1")
+	dut1Intf2 := dut1.Port(t, "port2")
+	dut1Intf3 := dut1.Port(t, "port3")
+	dut1Intf4 := dut1.Port(t, "port4")
+	rt := ondatra.ATE(t, "ate")
+	t.Logf("dut-1 %v", dut1)
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:837]: no need to log all the ports. Ondatra will print the reservation information, so this looks duplicative to me

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 15]
```diff
@@ -0,0 +1,1558 @@
...
+		for k, key := range keys {
+			VerifyLoadbalance(t, 4, p.trafficRate, p.trafficDuration, ftiIntfCount, int64(dut2InitialFitStats[key]), int64(dut2FinalFitStats[key]), loadbalance[k])
+		}
+	}
+}
+
+func disableAndEnableTunnelInterface(t *testing.T, dut *ondatra.DUTDevice, state bool) {
+
+	dutIntfs := []struct {
+		desc     string
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:1521]: the desc value is the same for all interfaces, so just inline the string "tunnel_endpoint" and simplify this to a slice of strings

**Review by greg-dennis — CHANGES_REQUESTED** [review]

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go (hunk 12 above)
Comment (greg-dennis) [reply feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:740]: Hmm, I see how other functions accept *parameters. Let me think about this one.

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 16]
```diff
@@ -0,0 +1,1558 @@
...
+	wantLoss = true
+	loadbalance = []bool{true, true, true, true, true, true, true, true}
+	t.Run("Verify load balance and traffic drops with IPv4 and IPv6 flow via agin 8 tunnel", func(t *testing.T) {
+		t.Log("Verify load balance and traffic drops with IPv4 and IPv6 flow via agin 8 tunnel")
+		VerifyUnderlayOverlayLoadbalanceTest(t, p, dut1, dut2, rt, dut1Intf1, dut1Intf2, dut1Intf3, dut1Intf4, dut2Intf1, dut2Intf2, dut2Intf3, dut2Intf4, 8, wantLoss, loadbalance)
+	})
+}
+
+func configureOtg(t *testing.T, otg *otg.OTG, p *parameters) gosnappi.Config {
+
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:1128]: delete blank line

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 17]
```diff
@@ -0,0 +1,1558 @@
...
+	iDut4Ipv4.SetAddress(p.rtIntf6Ipv4Add).SetGateway(p.r1Intf6Ipv4Add).SetPrefix(int32(p.ipv4Mask))
+	iDut4Ipv6 := iDut4Eth.Ipv6Addresses().Add().SetName("port6" + ".IPv6")
+	iDut4Ipv6.SetAddress(p.rtIntf6Ipv6Add).SetGateway(p.r1Intf6Ipv6Add).SetPrefix(int32(p.ipv6Mask))
+
+	t.Logf("Start Ote Traffic config")
+	t.Logf("configure IPv4 flow from %s to %s ", port1.Name(), port3.Name())
+	// Set config flow
+	flow1ipv4 := config.Flows().Add().SetName(p.flow1)
+	flow1ipv4.Metrics().SetEnable(true)
+	// Set source and reciving ports.
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:1178]: *receiving

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 18]
```diff
@@ -0,0 +1,1558 @@
...
+	loadbalance = []bool{true, true, true, true, true, true, true, true}
+	t.Run("Verify load balance and traffic drops with IPv4 and IPv6 flow via agin 8 tunnel", func(t *testing.T) {
+		t.Log("Verify load balance and traffic drops with IPv4 and IPv6 flow via agin 8 tunnel")
+		VerifyUnderlayOverlayLoadbalanceTest(t, p, dut1, dut2, rt, dut1Intf1, dut1Intf2, dut1Intf3, dut1Intf4, dut2Intf1, dut2Intf2, dut2Intf3, dut2Intf4, 8, wantLoss, loadbalance)
+	})
+}
+
+func configureOtg(t *testing.T, otg *otg.OTG, p *parameters) gosnappi.Config {
+
+	//  NewConfig creates a new OTG config.
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:1129]: this comment is obvious, delete

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 19]
```diff
@@ -0,0 +1,1558 @@
...
+	config := otg.NewConfig(t)
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:1130]: otg.NewConfig(t) is deprecated. Just use gosnappi.NewConfig()

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 20]
```diff
@@ -0,0 +1,1558 @@
...
+
+	Addr := net.ParseIP(address)
+	var network net.IP
+	_ = network
+	IsIPv4 := Addr.To4()
+	if IsIPv4 != nil {
+		// This mask corresponds to a /24 subnet for IPv4.
+
+		ipv4Mask := net.CIDRMask(mask, 32)
+		//t.Logf("%s in %T\n",ipv4Mask,ipv4Mask)
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:215]: delete line

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 21]
```diff
@@ -0,0 +1,1558 @@
...
+
+func GetNetworkAddress(t *testing.T, address string, mask int) string {
+
+	Addr := net.ParseIP(address)
+	var network net.IP
+	_ = network
+	IsIPv4 := Addr.To4()
+	if IsIPv4 != nil {
+		// This mask corresponds to a /24 subnet for IPv4.
+
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:213]: delete blank line

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 22]
```diff
@@ -0,0 +1,1558 @@
...
+		ipv4Mask := net.CIDRMask(mask, 32)
+		//t.Logf("%s in %T\n",ipv4Mask,ipv4Mask)
+		network := Addr.Mask(ipv4Mask)
+		net := fmt.Sprintf("%s/%d", network, mask)
+		t.Logf("network address : %s", net)
+		return net
+	} else {
+
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:221]: delete blank line

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 23]
```diff
@@ -0,0 +1,1558 @@
...
+		// This mask corresponds to a /32 subnet for IPv6.
+		ipv6Mask := net.CIDRMask(mask, 128)
+		network := Addr.Mask(ipv6Mask)
+		//t.Logf("IPv6 network: %s",network)
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:225]: delete line

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 24]
```diff
@@ -0,0 +1,1558 @@
...
+		net := fmt.Sprintf("%s/%d", network, mask)
+		t.Logf("Network address : %s", net)
+		return net
+	}
+
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:230]: delete blank line

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 25]
```diff
@@ -0,0 +1,1558 @@
...
+}
+
+func ConfigureTunnelEncapDUT(t *testing.T, p *parameters, dut *ondatra.DUTDevice, dp1 *ondatra.Port, dp2 *ondatra.Port, dp3 *ondatra.Port, dp4 *ondatra.Port) {
+
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:234]: delete blank line

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 26]
```diff
@@ -0,0 +1,1558 @@
...
+		network := Addr.Mask(ipv6Mask)
+		//t.Logf("IPv6 network: %s",network)
+		net := fmt.Sprintf("%s/%d", network, mask)
+		t.Logf("Network address : %s", net)
+		return net
+	}
+
+}
+
+func ConfigureTunnelEncapDUT(t *testing.T, p *parameters, dut *ondatra.DUTDevice, dp1 *ondatra.Port, dp2 *ondatra.Port, dp3 *ondatra.Port, dp4 *ondatra.Port) {
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:233]: no need for this method to be exported -- make it start with lowercase

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 27]
```diff
@@ -0,0 +1,1558 @@
...
+
+		// configure ipv6 address
+		i6 := i.GetOrCreateSubinterface(0).GetOrCreateIpv6()
+		b := i6.GetOrCreateAddress(intf.ipv6Addr)
+		b.PrefixLength = ygot.Uint8(intf.ipv6mask)
+		gnmi.Replace(t, dut, gnmi.OC().Interface(intf.intfName).Config(), i)
+	}
+}
+
+func ConfigureTunnelDecapDUT(t *testing.T, p *parameters, dut *ondatra.DUTDevice, dp1 *ondatra.Port, dp2 *ondatra.Port, dp3 *ondatra.Port, dp4 *ondatra.Port) {
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:375]: no need to export

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 28]
```diff
@@ -0,0 +1,1558 @@
...
+
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:376]: delete blank line

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 29]
```diff
@@ -0,0 +1,1558 @@
...
+		// configure ipv6 address
+		i6 := i.GetOrCreateSubinterface(0).GetOrCreateIpv6()
+		b := i6.GetOrCreateAddress(intf.ipv6Addr)
+		b.PrefixLength = ygot.Uint8(intf.ipv6mask)
+		gnmi.Replace(t, dut, gnmi.OC().Interface(intf.intfName).Config(), i)
+	}
+}
+
+func ConfigureTunnelInterface(t *testing.T, intf string, tunnelSrc string, tunnelDst string, dut *ondatra.DUTDevice) {
+
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:519]: delete blank line

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 30]
```diff
@@ -0,0 +1,1558 @@
...
+	// IPv4 tunnel source and destination configuration
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:520]: All single-line comments need to end with punctuation, per go style

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 31]
```diff
@@ -0,0 +1,1558 @@
...
+		if li == lj {
+			return idi < idj
+		}
+		return li < lj // "port2" < "port10"
+	})
+	return ports
+}
+
+func ConfigureTunnelEndPoints(intf string, tunnelSrc string, tunnelDest string) string {
+
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:555]: delete blank line

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 32]
```diff
@@ -0,0 +1,1558 @@
...
+		li, lj := len(idi), len(idj)
+		if li == lj {
+			return idi < idj
+		}
+		return li < lj // "port2" < "port10"
+	})
+	return ports
+}
+
+func ConfigureTunnelEndPoints(intf string, tunnelSrc string, tunnelDest string) string {
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:554]: no need to export, same with all other helper methods throughout

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 33]
```diff
@@ -0,0 +1,1558 @@
...
+                address %s;
+            }
+        }
+    }
+}`, address)
+
+}
+
+func ConfigureAdditionalIPv6AddressonLoopback(address string) string {
+
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:593]: delete blank line

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 34]
```diff
@@ -0,0 +1,1558 @@
...
+    lo0 {
+        unit 0 {
+            family inet6 {
+                address %s;
+            }
+        }
+    }
+}`, address)
+
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:605]: delete blank line

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 35]
```diff
@@ -0,0 +1,1558 @@
...
+}
+
+func ConfigureTunnelTerminationOption(interf string) string {
+
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:609]: delete blank line, and so on for all other "first lines of methods" that are blank

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 36]
```diff
@@ -0,0 +1,1558 @@
...
+	t.Logf("dut2-port3 %v", dut2Intf3)
+	t.Logf("dut2-port3 %v", dut2Intf4)
+
+	t.Run("Configure dut1 and dut2 ", func(t *testing.T) {
+		ConfigureTunnelEncapDUT(t, p, dut1, dut1Intf1, dut1Intf2, dut1Intf3, dut1Intf4)
+		ConfigureTunnelDecapDUT(t, p, dut2, dut2Intf1, dut2Intf2, dut2Intf3, dut2Intf4)
+	})
+
+	t.Run("Configure loopback interface on dut1 and dut2 ", func(t *testing.T) {
+		// configure addtional loop address by native cli configuration.
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:861]: All single-line comments need to start uppercase. Same throughout.

File: feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go [hunk 37]
```diff
@@ -0,0 +1,1558 @@
...
+		ConfigureTunnelInterface(t, "fti2", p.r0Lo0Ut2Ipv4Add, p.r1Lo0Ut2Ipv4Add, dut1)
+		ConfigureTunnelInterface(t, "fti3", p.r0Lo0Ut3Ipv4Add, p.r1Lo0Ut3Ipv4Add, dut1)
+
+		// configure tunnel interface on dut2- IPv4
+		ConfigureTunnelInterface(t, "fti0", p.r1Lo0Ut0Ipv4Add, p.r0Lo0Ut0Ipv4Add, dut2)
+		ConfigureTunnelInterface(t, "fti1", p.r1Lo0Ut1Ipv4Add, p.r0Lo0Ut1Ipv4Add, dut2)
+		ConfigureTunnelInterface(t, "fti2", p.r1Lo0Ut2Ipv4Add, p.r0Lo0Ut2Ipv4Add, dut2)
+		ConfigureTunnelInterface(t, "fti3", p.r1Lo0Ut3Ipv4Add, p.r0Lo0Ut3Ipv4Add, dut2)
+
+		//configure tunnel interface on dut2- IPv6
```
Comment (greg-dennis) [inline feature/experimental/tunnel/otg_tests/tunnel_interface_based_disable_enable_test/tunnel_interface_disable_enable_test.go:892]: All line comments need a space after //

**Comment by sachendras** [comment]
This test is now deprecated as it is covered by other tunnel tests.

**Pull request closed without merging**